pkg/deploy/deploy.go

```diff
@@ -158,20 +159,35 @@ func (d *deployer) deploy(ctx context.Context, rgName, deploymentName, vmssName
 	for i := 0; i < 3; i++ {
 		d.log.Printf("deploying %s", deploymentName)
 		err = d.deployments.CreateOrUpdateAndWait(ctx, rgName, deploymentName, deployment)
-		if serviceErr, ok := err.(*azure.ServiceError); ok &&
-			serviceErr.Code == "DeploymentFailed" &&
-			i < 1 {
-			// on new RP deployments, we get a spurious DeploymentFailed error
+		serviceErr, isServiceError := err.(*azure.ServiceError)
+
+		// Check for a known error that we know how to handle.
+		if isServiceError && serviceErr.Code == "DeploymentFailed" && len(serviceErr.Details) > 0 {
+			outerErr := azure.ServiceError{}
+			jsonEncoded, _ := json.Marshal(serviceErr.Details[0])
```

Comment:
can we check the error here?

Reply:
I intentionally left it out to keep things simple because if we get to the conditional where we check the error contents and there have been problems with the JSON marshaling/unmarshaling, `outerErr` and `innerErr`'s fields will either be empty or not contain what we're looking for. Do you think it's still worth checking it?

Reply:
IMO this json marshalling/unmarshalling should be in a helper function where we can have unit tests to confirm which error handling strategy we want to go with.

Reply:
I've moved that functionality to a helper function. I'm not 100% confident in the way I handled the errors it returns though, so let me know what you think about it.

Reply:
I think you did a fine job, there is still a bit of nesting that could be removed, but that's more of a nitpick

Reply:
Yeah, I see what you mean. I found one spot where I removed some nesting.